flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackend.java

```diff
@@ -172,4 +174,19 @@ OperatorStateBackend createOperatorStateBackend(
     default boolean useManagedMemory() {
         return false;
     }
+
+    /**
+     * Tells if a state backend supports taking forced full snapshots. Forced full snapshots are
+     * snapshots that do not share files with previous snapshots. The feature is necessary to
+     * properly support {@link RestoreMode#NO_CLAIM} mode.
+     *
+     * <p>If a state backend supports forced full snapshots, it should create an independent
+     * snapshot when it receives {@link CheckpointOptions#isForcedFullSnapshot()} request in {@link
+     * Snapshotable#snapshot(long, long, CheckpointStreamFactory, CheckpointOptions)}.
+     *
+     * @return If the state backend supports taking forced full snapshots.
+     */
+    default boolean supportsForcedFullSnapshot() {
```

Comment:
Hey @rkhachatryan @Myasuka What do you think about this approach? This allows us to have a migration path for state backends that do not support forced full snapshots yet. We would add support to all built-in state backends, but it would work also for external state backends.

WDYT?

Reply:
I think it's okay to tell Flink whether this state backend supports to run force full snapshot.
However, shall we distinguish keyed state backend and operator state backend here?

Reply:
It does not need to do a reupload. It will duplicate files in the future. Also in case of other backends which do not support incremental snapshots, they do not need to do anything in such cases. The intention is to take independent/full snapshot.

Reply:
I'd consider duplication as an optimization.
Anyways, I think the idea is to create not a full checkpoint, but to "break the lineage". So independent and full are distinct to me. For example, it could be enough to re-upload/duplicate the files added before some intial checkpoint.

Btw, what do you think about adding checkpointID as an argument, so that the above optimization (re-upload only old) can be implemented?

Reply:
I am not strong on that naming. I don't like the `supportsForceReupload`, imo it describes it even worse. Any other ideas?

As for the `checkpointId`, what would it mean in that context? The purpose of this method is to add a feature flag to a state backend. We either do or we do not support the feature. Not sure how `checkpointID` can be helpful in here. 

Theoretically speaking the flag my arrive at any point in time. I don't think it would work nicely to fail e.g. after n checkpoints safely taken.

Reply:
In the course of working on passing the flag through for "forced full checkpoint" I thought I could name this particular function `supportsNoClaimRestoreMode`. WDYT about that proposal?

Reply:
> As for the checkpointId, what would it mean in that context?

My reasoning is that we don't have to re-upload **all** state if first `N` checkpoints fail, only the "old" state. 
And I was erroneously assuming that there will be an explicit method in StateBackend to take such a snapshot. In reality, `SnapshotStrategy` analyzes `CheckpointOptions`.
Passing this `initialCheckpointID` in `CheckpointOptions` seems too fragile and complex.
Sorry for confusion.

`supportsNoClaimRestoreMode` sounds better :+1: